C compilation error (from tcc):

### Describe the bug

C compilation error (from tcc):

### Reproduction Steps

```
module main

import db.mysql
import db.pg

type DBconfig = mysql.Config | pg.Config

struct ConnectionPoolGeneric[T] {
mut:
	connections chan T   // mysql.DB, pg.DB
	config      DBconfig // mysql.Config, pg.Config
}

pub fn new_conn_pool[T](conf DBconfig, size int) !&ConnectionPoolGeneric[T] {
	$if conf is mysql.Config {
		mysql_conf := conf as mysql.Config
		mut pool := &ConnectionPoolGeneric[mysql.DB]{
			connections: chan mysql.DB{cap: size}
			config:      mysql_conf
		}
		for _ in 0 .. size {
			conn := mysql.connect(mysql_conf)!
			poo.connections <- conn
		}
		return pool
	} $else $if conf is pg.Config {
		pg_conf := conf as pg.Config
		mut pool := &ConnectionPoolGeneric[pg.DB]{
			connections: chan pg.DB{cap: size}
			config:      pg_conf
		}
		for _ in 0 .. size {
			conn := pg.connect(pg_conf)!
			pool.connections <- conn
		}
		return pool
	}$else{
	  return error('')
	}
}

// acquire gets a connection from the pool | 从池中获取连接
pub fn (mut pool ConnectionPoolGeneric[T]) acquire() !T {
	conn := <-pool.connections or { return error('Failed mysql') }
	return conn
}

// release returns a connection back to the pool. | 释放将连接返回到池中
pub fn (mut pool ConnectionPoolGeneric[T]) release(conn T) {
	pool.connections <- conn
}

// close closes all connections in the pool | 关闭池中的所有连接
pub fn (mut pool ConnectionPoolGeneric[T]) close() {
	for _ in 0 .. pool.connections.len {
		mut conn := <-pool.connections or { break }
		conn.close()
	}
}


const config = mysql.Config{
	host:     '127.0.0.1'
	port:     3306
	username: 'root'
	password: 'mysql_123456'
	dbname:   'vcore'
}

fn main() {
	mut pool := new_conn_pool[mysql.DB](config, 5) or {
		panic('Failed to create mysql pool: ${err}')
	}
	dump(pool)
	db := pool.acquire() or { panic(err) }
	dump(db)
	data := db.exec('select 1')!
	dump(data)
	assert data[0] == mysql.Row{
		vals: ['1']
	}
	pool.release(db)
	pool.close()
}

```
v run pub_pool_db.v


### Expected Behavior

success

### Current Behavior

```
Jengro@Jengro-PC:~/Documents/Dev/v-product/vprod-workspase/backend/common/dbpool$ v run pub_pool_db.v
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/pub_pool_db.01JVVTYN2KREEJD5NBYZ1B2B7S.tmp.c:15216: warning: implicit declaration of function 'db__mysql__DB_close'
cc: /home/Jengro/.vmr/versions/v_versions/v_latest/thirdparty/tcc/lib/libgc.a: error: 'GC_noop1_ptr' defined twice
cc: tcc: error: undefined symbol 'db__mysql__DB_close'
...
cc: tcc: error: undefined symbol 'db__mysql__DB_close'
(note: the original output was 6 lines long; it was truncated to its first 3 lines + the last line)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.10 299ebdf

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.10 2c42574e6a25b4c5ca8381a63a05a0d0b997436f.299ebdf
|:-------------------|:-------------------
|OS                  |linux, Deepin 23.1
|Processor           |6 cpus, 64bit, little endian, Intel(R) Core(TM) i5-8500B CPU @ 3.00GHz
|Memory              |0.22GB/7.61GB
|                    |
|V executable        |/home/Jengro/.vmr/versions/v_versions/v_latest/v
|V last modified time|2025-05-22 11:09:17
|                    |
|V home dir          |OK, value: /home/Jengro/.vmr/versions/v_versions/v_latest
|VMODULES            |OK, value: /home/Jengro/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/Jengro/Documents/Dev/v-product/vprod-workspase/backend/common/dbpool
|                    |
|Git version         |git version 2.47.2
|V git status        |weekly.2025.20-50-g299ebdff
|.git/config present |true
|                    |
|cc version          |cc (Deepin 12.3.0-17deepin12) 12.3.0
|gcc version         |gcc (Deepin 12.3.0-17deepin12) 12.3.0
|clang version       |Deepin clang version 17.0.6 (5deepin5)
|tcc version         |tcc version 0.9.28rc 2025-02-13 HEAD@f8bd136d (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 696c1d84
|emcc version        |N/A
|glibc version       |ldd (Debian GLIBC 2.38-6deepin13+rb1) 2.38
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

Using `-no-skip-unused` solves some failures, but still others:

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/bug.01JVVWDTFGV0PPB3HBSVJX6KN6.tmp.c:26187: warning: assignment discards qualifiers from pointer target type
cc: /tmp/v_1000/bug.01JVVWDTFGV0PPB3HBSVJX6KN6.tmp.c:27121: error: cannot convert 'struct db__mysql__DB' to 'struct db__mysql__DB *'
...
cc: /tmp/v_1000/bug.01JVVWDTFGV0PPB3HBSVJX6KN6.tmp.c:27121: error: cannot convert 'struct db__mysql__DB' to 'struct db__mysql__DB *'
(note: the original output was 9 lines long; it was truncated to its first 2 lines + the last line)
=================================================================
```

--- a/vlib/db/pg/pg.c.v
+++ b/vlib/db/pg/pg.c.v
@@ -224,7 +224,7 @@ fn res_to_rows(res voidptr) []Row {
 }
 
 // close frees the underlying resource allocated by the database connection
-pub fn (db DB) close() {
+pub fn (db &DB) close() {
 	C.PQfinish(db.conn)
 }
 
@@ -232,7 +232,7 @@ pub fn (db DB) close() {
 // returns an the first field in the first tuple
 // converted to an int. If no row is found or on
 // command failure, an error is returned
-pub fn (db DB) q_int(query string) !int {
+pub fn (db &DB) q_int(query string) !int {
 	rows := db.exec(query)!
 	if rows.len == 0 {
 		return error('q_int "${query}" not found')
@@ -249,7 +249,7 @@ pub fn (db DB) q_int(query string) !int {
 // returns an the first field in the first tuple
 // as a string. If no row is found or on
 // command failure, an error is returned
-pub fn (db DB) q_string(query string) !string {
+pub fn (db &DB) q_string(query string) !string {
 	rows := db.exec(query)!
 	if rows.len == 0 {
 		return error('q_string "${query}" not found')
@@ -264,12 +264,12 @@ pub fn (db DB) q_string(query string) !string {
 
 // q_strings submit a command to the database server and
 // returns the resulting row set. Alias of `exec`
-pub fn (db DB) q_strings(query string) ![]Row {
+pub fn (db &DB) q_strings(query string) ![]Row {
 	return db.exec(query)
 }
 
 // exec submits a command to the database server and wait for the result, returning an error on failure and a row set on success
-pub fn (db DB) exec(query string) ![]Row {
+pub fn (db &DB) exec(query string) ![]Row {
 	res := C.PQexec(db.conn, &char(query.str))
 	return db.handle_error_or_result(res, 'exec')
 }
@@ -282,7 +282,7 @@ fn rows_first_or_empty(rows []Row) !Row {
 }
 
 // exec_one executes a query and returns its first row as a result, or an error on failure
-pub fn (db DB) exec_one(query string) !Row {
+pub fn (db &DB) exec_one(query string) !Row {
 	res := C.PQexec(db.conn, &char(query.str))
 	e := unsafe { C.PQerrorMessage(db.conn).vstring() }
 	if e != '' {
@@ -293,7 +293,7 @@ pub fn (db DB) exec_one(query string) !Row {
 }
 
 // exec_param_many executes a query with the parameters provided as ($1), ($2), ($n)
-pub fn (db DB) exec_param_many(query string, params []string) ![]Row {
+pub fn (db &DB) exec_param_many(query string, params []string) ![]Row {
 	unsafe {
 		mut param_vals := []&char{len: params.len}
 		for i in 0 .. params.len {
@@ -307,24 +307,24 @@ pub fn (db DB) exec_param_many(query string, params []string) ![]Row {
 }
 
 // exec_param executes a query with 1 parameter ($1), and returns either an error on failure, or the full result set on success
-pub fn (db DB) exec_param(query string, param string) ![]Row {
+pub fn (db &DB) exec_param(query string, param string) ![]Row {
 	return db.exec_param_many(query, [param])
 }
 
 // exec_param2 executes a query with 2 parameters ($1) and ($2), and returns either an error on failure, or the full result set on success
-pub fn (db DB) exec_param2(query string, param string, param2 string) ![]Row {
+pub fn (db &DB) exec_param2(query string, param string, param2 string) ![]Row {
 	return db.exec_param_many(query, [param, param2])
 }
 
 // prepare submits a request to create a prepared statement with the given parameters, and waits for completion. You must provide the number of parameters (`$1, $2, $3 ...`) used in the statement
-pub fn (db DB) prepare(name string, query string, num_params int) ! {
+pub fn (db &DB) prepare(name string, query string, num_params int) ! {
 	res := C.PQprepare(db.conn, &char(name.str), &char(query.str), num_params, 0) // defining param types is optional
 
 	return db.handle_error(res, 'prepare')
 }
 
 // exec_prepared sends a request to execute a prepared statement with given parameters, and waits for the result. The number of parameters must match with the parameters declared in the prepared statement.
-pub fn (db DB) exec_prepared(name string, params []string) ![]Row {
+pub fn (db &DB) exec_prepared(name string, params []string) ![]Row {
 	unsafe {
 		mut param_vals := []&char{len: params.len}
 		for i in 0 .. params.len {
@@ -337,7 +337,7 @@ pub fn (db DB) exec_prepared(name string, params []string) ![]Row {
 	}
 }
 
-fn (db DB) handle_error_or_result(res voidptr, elabel string) ![]Row {
+fn (db &DB) handle_error_or_result(res voidptr, elabel string) ![]Row {
 	e := unsafe { C.PQerrorMessage(db.conn).vstring() }
 	if e != '' {
 		C.PQclear(res)
@@ -349,7 +349,7 @@ fn (db DB) handle_error_or_result(res voidptr, elabel string) ![]Row {
 	return res_to_rows(res)
 }
 
-fn (db DB) handle_error(res voidptr, elabel string) ! {
+fn (db &DB) handle_error(res voidptr, elabel string) ! {
 	e := unsafe { C.PQerrorMessage(db.conn).vstring() }
 	if e != '' {
 		C.PQclear(res)
@@ -362,7 +362,7 @@ fn (db DB) handle_error(res voidptr, elabel string) ! {
 
 // copy_expert executes COPY command
 // https://www.postgresql.org/docs/9.5/libpq-copy.html
-pub fn (db DB) copy_expert(query string, mut file io.ReaderWriter) !int {
+pub fn (db &DB) copy_expert(query string, mut file io.ReaderWriter) !int {
 	mut res := C.PQexec(db.conn, &char(query.str))
 	status := unsafe { ExecStatusType(C.PQresultStatus(res)) }
 	defer {
@@ -423,7 +423,7 @@ pub fn (db DB) copy_expert(query string, mut file io.ReaderWriter) !int {
 	return 0
 }
 
-fn pg_stmt_worker(db DB, query string, data orm.QueryData, where orm.QueryData) ![]Row {
+fn pg_stmt_worker(db &DB, query string, data orm.QueryData, where orm.QueryData) ![]Row {
 	mut param_types := []u32{}
 	mut param_vals := []&char{}
 	mut param_lens := []int{}
@@ -452,7 +452,7 @@ pub struct PQTransactionParam {
 }
 
 // begin begins a new transaction.
-pub fn (db DB) begin(param PQTransactionParam) ! {
+pub fn (db &DB) begin(param PQTransactionParam) ! {
 	mut sql_stmt := 'BEGIN TRANSACTION ISOLATION LEVEL '
 	match param.transaction_level {
 		.read_uncommitted { sql_stmt += 'READ UNCOMMITTED' }
@@ -468,7 +468,7 @@ pub fn (db DB) begin(param PQTransactionParam) ! {
 }
 
 // commit commits the current transaction.
-pub fn (db DB) commit() ! {
+pub fn (db &DB) commit() ! {
 	_ := C.PQexec(db.conn, c'COMMIT;')
 	e := unsafe { C.PQerrorMessage(db.conn).vstring() }
 	if e != '' {
@@ -477,7 +477,7 @@ pub fn (db DB) commit() ! {
 }
 
 // rollback rollbacks the current transaction.
-pub fn (db DB) rollback() ! {
+pub fn (db &DB) rollback() ! {
 	_ := C.PQexec(db.conn, c'ROLLBACK;')
 	e := unsafe { C.PQerrorMessage(db.conn).vstring() }
 	if e != '' {
@@ -486,7 +486,7 @@ pub fn (db DB) rollback() ! {
 }
 
 // rollback_to rollbacks to a specified savepoint.
-pub fn (db DB) rollback_to(savepoint string) ! {
+pub fn (db &DB) rollback_to(savepoint string) ! {
 	if !savepoint.is_identifier() {
 		return error('savepoint should be a identifier string')
 	}
@@ -499,7 +499,7 @@ pub fn (db DB) rollback_to(savepoint string) ! {
 }
 
 // savepoint create a new savepoint.
-pub fn (db DB) savepoint(savepoint string) ! {
+pub fn (db &DB) savepoint(savepoint string) ! {
 	if !savepoint.is_identifier() {
 		return error('savepoint should be a identifier string')
 	}

--- a/vlib/db/pg/pg_double_test.v
+++ b/vlib/db/pg/pg_double_test.v
@@ -16,7 +16,7 @@ fn test_float_field() {
 		eprintln('> This test requires a working postgres server running on localhost.')
 		return
 	}
-	conn := 'host=localhost user=test password=test' // insert own connection string
+	conn := 'host=localhost user=postgres password=12345678' // insert own connection string
 	db := pg.connect_with_conninfo(conn)!
 	defer {
 		db.close()

--- a/vlib/db/pg/pg_orm_test.v
+++ b/vlib/db/pg/pg_orm_test.v
@@ -1,4 +1,3 @@
-// vtest build: started_postgres?
 module main
 
 import orm
@@ -45,13 +44,14 @@ fn test_pg_orm() {
 	mut db := pg.connect(
 		host:     'localhost'
 		user:     'postgres'
-		password: 'password'
-		dbname:   'postgres'
+		password: '12345678'
+		dbname:   'test'
 	) or { panic(err) }
 
 	defer {
 		db.close()
 	}
+	db.drop('Test')!
 
 	db.create('Test', [
 		orm.TableField{
@@ -140,6 +140,9 @@ fn test_pg_orm() {
 	/** test orm sql type
 	* - verify if all type create by attribute sql_type has created
 	*/
+	sql db {
+		drop table TestCustomSqlType
+	}!
 
 	sql db {
 		create table TestCustomSqlType
@@ -157,14 +160,11 @@ fn test_pg_orm() {
 	mut information_schema_data_types_results := []string{}
 	information_schema_custom_sql := ['integer', 'text', 'character varying',
 		'timestamp without time zone', 'uuid']
+
 	for data_type in result_custom_sql {
-		x := data_type.vals[0]!
+		x := data_type.vals[0]
 		information_schema_data_types_results << x?
 	}
-
-	sql db {
-		drop table TestCustomSqlType
-	}!
 
 	assert information_schema_data_types_results == information_schema_custom_sql
 
@@ -206,6 +206,9 @@ fn test_pg_orm() {
 	assert results[0].updated_at == model.updated_at
 	assert results[0].deleted_at == model.deleted_at
 
+	sql db {
+		drop table TestDefaultAttribute
+	} or {}
 	/** test default attribute
 	*/
 	sql db {
@@ -224,8 +227,8 @@ fn test_pg_orm() {
 	mut information_schema_defaults_results := []string{}
 
 	for defaults in result_defaults {
-		x := defaults.vals[0]!
-		information_schema_defaults_results << x?
+		x := defaults.vals[0]
+		information_schema_defaults_results << x or { '' }
 	}
 	sql db {
 		drop table TestDefaultAttribute

--- a/vlib/db/pg/pg_orm_test.v
+++ b/vlib/db/pg/pg_orm_test.v
@@ -1,3 +1,4 @@
+// vtest build: started_postgres?
 module main
 
 import orm
@@ -206,9 +207,6 @@ fn test_pg_orm() {
 	assert results[0].updated_at == model.updated_at
 	assert results[0].deleted_at == model.deleted_at
 
-	sql db {
-		drop table TestDefaultAttribute
-	} or {}
 	/** test default attribute
 	*/
 	sql db {

